b2sdk/transfer/outbound/upload_source.py

```diff
@@ -126,26 +141,43 @@ def __repr__(self):
             id=id(self),
         )
 
-    def get_content_length(self):
+    def get_content_length(self) -> int:
         return self.content_length
 
-    def get_content_sha1(self):
+    def _get_hexdigest(self, length: int = 0) -> Sha1HexDigest:
+        length = length or self.content_length
+
+        if self.digest is None:
+            self.digest = hashlib.sha1()
+            self.digest_progress = 0
+
+        if length < self.digest_progress:
```

Comment:
if we want to compare an upload source to a potential incremental target, we can only do so once. Later this UploadSourceLocalFile will crash. I think this is wrong.

There should be an IncrementalHexDigester object that UploadSourceLocalFile can manufacture for you and then _that_ object keeps the digest_progress state, not UploadSourceLocalFile (which is a read-only thing essentially, other than the manually cached property for the sha1 of the entire file)

Reply:
I mean... Thus far this class has been threadsafe. Not documented as such, but it was safe. Now you were making it stateful, so someone who'd inherit from `UploadSourceLocalFileBase` might be surprised that `self.get_content_sha1` is non-threadsafe and `self._get_hexdigest()` is stateful, so it'll crash frequently when run with multiple threads.

With the proposed change you make it retryable, so one thread can call it several times and it won't crash, but still multiple threads will have issues.

We could document this class as non-threadsafe, but I'd rather keep the undocumented state and make it stateless and threadsafe.

Reply:
Fair point. I'll check whether I can make it stateless, nice and not loading the same file over and over again. If not, I'll just make it stateless.